eth/block-validation/api_test.go

```diff
@@ -870,3 +872,310 @@ func TestBlockAssemblerRequestJsonEncodingAndDecoding(t *testing.T) {
+	})
+	fmt.Printf("DEBUG: Exec data random is %x\n", execData.Random)
+	require.NoError(t, err)
+	require.EqualValues(t, len(execData.Withdrawals), 2)
+	require.EqualValues(t, len(execData.Transactions), 0)
+
+	payload, err := ExecutableDataToExecutionPayloadV2(execData)
+	require.NoError(t, err)
+
+	proposerAddr := bellatrix.ExecutionAddress{}
+	copy(proposerAddr[:], testValidatorAddr.Bytes())
+
+	blockRequest := &BuilderBlockValidationRequestV2{
+		SubmitBlockRequest: capellaapi.SubmitBlockRequest{
+			Signature: phase0.BLSSignature{},
+			Message: &apiv1.BidTrace{
+				ParentHash:           phase0.Hash32(execData.ParentHash),
+				BlockHash:            phase0.Hash32(execData.BlockHash),
+				ProposerFeeRecipient: proposerAddr,
+				GasLimit:             execData.GasLimit,
+				GasUsed:              execData.GasUsed,
+				// This value is actual profit + 1, validation should fail
+				Value: uint256.NewInt(149842511727213),
+			},
+			ExecutionPayload: payload,
+		},
+		RegisteredGasLimit: execData.GasLimit,
+		WithdrawalsRoot:    withdrawalsRoot,
+	}
+
+	assemblyRequest := BlockAssemblerRequest{
+		TobTxs:             tobTxs,
+		RobPayload:         blockRequest.SubmitBlockRequest,
+		RegisteredGasLimit: execData.GasLimit,
+	}
+
+	block, err := api.BlockAssembler(&assemblyRequest)
+	require.NoError(t, err)
+	require.Equal(t, len(block.Transactions), 2)
+
+	// check tob txs
+	actualTobTx1 := block.Transactions[0]
+	actualTobTx2 := block.Transactions[1]
+	require.Equal(t, bellatrixUtil.ExecutionPayloadTransactions{Transactions: []bellatrix.Transaction{actualTobTx1, actualTobTx2}}, tobTxs)
+}
+
+func TestBlockAssemblerWithNoTobTxs(t *testing.T) {
+	genesis, preMergeBlocks := generatePreMergeChain(20)
+	os.Setenv("BUILDER_TX_SIGNING_KEY", testBuilderKeyHex)
+	time := preMergeBlocks[len(preMergeBlocks)-1].Time() + 5
+	genesis.Config.ShanghaiTime = &time
+
+	n, ethservice := startEthService(t, genesis, preMergeBlocks)
+	ethservice.Merger().ReachTTD()
+	defer n.Close()
+
+	api := NewBlockValidationAPI(ethservice, nil, false)
+	parent := preMergeBlocks[len(preMergeBlocks)-1]
+
+	api.eth.APIBackend.Miner().SetEtherbase(testBuilderAddr)
+
+	statedb, _ := ethservice.BlockChain().StateAt(parent.Root())
+	nonce := statedb.GetNonce(testAddr)
+
+	signer := types.LatestSigner(ethservice.BlockChain().Config())
+
+	robTx1, _ := types.SignTx(types.NewTransaction(nonce, common.Address{0x18}, big.NewInt(30), 21000, big.NewInt(2*params.InitialBaseFee), nil), signer, testKey)
+	robTx2, _ := types.SignTx(types.NewTransaction(nonce, common.Address{0x19}, big.NewInt(40), 21000, big.NewInt(2*params.InitialBaseFee), nil), signer, testKey)
+
+	ethservice.TxPool().AddLocal(robTx1)
+	ethservice.TxPool().AddLocal(robTx2)
+
+	withdrawals := []*types.Withdrawal{
+		{
+			Index:     0,
+			Validator: 1,
+			Amount:    100,
+			Address:   testAddr,
+		},
+		{
+			Index:     1,
+			Validator: 1,
+			Amount:    100,
+			Address:   testAddr,
+		},
```

Comment:
second use of this big var. could move to top of file?

Reply:
yes will refactor the tests